### PR TITLE
Partial progress to new logging system

### PR DESCRIPTION
**Pull request by Morlic-fo**
~~* Renaming all the previously existing ```info``` variables to allow the respective logger usage~~
* Use ```warn``` level logging instead of ```print >> sys.stderr``` for caught exceptions/inconsistencies which let the AI continue with its gameplay,

### CODE REVIEW

**Comment by Morlic-fo** [comment]
Rebased.

**Review by Dilvish-fo — APPROVED** [review]
Your proposed significance for warn() sounds fine to me; I think there are some existing uses of error() that should be swapped over to warn() then.

**Comment by Morlic-fo** [comment]
I added another 2 commits to convert current ```error``` level logs to ```warn``` level. I found only three existing use-cases that seem worthy to switch. Most of the ```error``` calls are in ```FreeorionAI.py``` and are done at fairly high-level function calls / exception handling. To make my previous statement more precise, ```warn``` level should probably only apply to fairly local/function level stuff where errors do not affect how the rest of the AI works.

For example,
```
    if not saved_state_string:
        error("AI given empty state-string to resume from; this is expected if the AI is assigned to an empire "
              "previously run by a human, but is otherwise an error. AI will be set to Aggressive.")
```
is technically handled so that the AI can play perfectly well but still seems significant enough to log at ```error``` level.

File: default/python/AI/AIstate.py [hunk 1]
```diff
@@ -182,9 +181,8 @@ def __setstate__(self, state):
             else:
                 try:
                     aggression = state['character'].get_trait(Aggression).key
-                except Exception as e:
-                    print >> sys.stderr, "Could not find the aggression level of the AI, defaulting to typical."
-                    print >> sys.stderr, e
+                except Exception:
+                    warn("Could not find the aggression level of the AI, defaulting to typical.", exc_info=True)
```
Comment (LGM-Doyle) [inline default/python/AI/AIstate.py:185]: This should be an error.  `character` and `aggression` are required and expected aspects of the AI model.  If `character` is missing from `state`, or `Aggression` is missing from `character` or `Aggression` has no key, these are all errors.  Furthermore, this is an unexpected `Exception`, so there will be traceback information suggesting it is more important than a warning.

File: default/python/AI/PlanetUtilsAI.py [hunk 2]
```diff
@@ -43,7 +41,7 @@ def get_capital():
     universe = fo.getUniverse()
     empire = fo.getEmpire()
     if empire is None:
-        print >> sys.stderr, "Danger Danger! FO can't find an empire for me!!!!"
+        warn("Danger Danger! FO can't find an empire for me!!!!")
```
Comment (LGM-Doyle) [inline default/python/AI/PlanetUtilsAI.py:44]: Danger Danger, but not an error?

File: default/python/AI/ShipDesignAI.py [hunk 3]
```diff
@@ -189,7 +187,7 @@ def print_hulls_for_planets(self, pid=None):
         elif isinstance(pid, list):
             planets = pid
         else:
-            print >> sys.stderr, "Invalid parameter 'pid' for 'print_hulls_for_planets'. Expected int, list or None."
+            warn("Invalid parameter 'pid' for 'print_hulls_for_planets'. Expected int, list or None.")
```
Comment (LGM-Doyle) [inline default/python/AI/ShipDesignAI.py:190]: This should be an error.  The type of `pid `is wrong, which should never happen.

File: default/python/AI/ShipDesignAI.py [hunk 4]
```diff
@@ -208,7 +206,7 @@ def print_parts_for_planets(self, pid=None):
         elif isinstance(pid, list):
             planets = pid
         else:
-            print >> sys.stderr, "Invalid parameter 'pid' for 'print_parts_for_planets'. Expected int, list or None."
+            warn("Invalid parameter 'pid' for 'print_parts_for_planets'. Expected int, list or None.")
```
Comment (LGM-Doyle) [inline default/python/AI/ShipDesignAI.py:209]: This should be an error.  The type of `pid `is wrong, which should never happen.

File: default/python/AI/ShipDesignAI.py [hunk 5]
```diff
@@ -372,17 +370,15 @@ def _check_cache_for_consistency(self):
             try:
                 cached_name = fo.getShipDesign(self.design_id_by_name[designname]).name
                 if cached_name != designname:
-                    print "  WARNING: ShipID cache corrupted."
-                    print "    Expected: %s, got: %s. Repairing cache." % (designname, cached_name)
+                    warn("ShipID cache corrupted. Expected: %s, got: %s." % (designname, cached_name))
```
Comment (LGM-Doyle) [inline default/python/AI/ShipDesignAI.py:373]: This should be an error.  The cache is broken, which should never happen, even with bad data.

File: default/python/AI/ShipDesignAI.py [hunk 6]
```diff
@@ -372,17 +370,15 @@ def _check_cache_for_consistency(self):
...
                     design_id = next(iter([shipDesignID for shipDesignID in fo.getEmpire().allShipDesigns
                                           if designname == fo.getShipDesign(shipDesignID).name]), None)
                     if design_id is not None:
                         self.design_id_by_name[designname] = design_id
                     else:
                         corrupted.append(designname)
             except AttributeError:
-                print "  WARNING: ShipID cache corrupted. Could not get cached shipdesign. Repairing Cache."
-                print traceback.format_exc()  # do not print to stderr as this is an "expected" exception.
+                warn("ShipID cache corrupted. Could not get cached shipdesign. Repairing Cache.", exc_info=True)
```
Comment (LGM-Doyle) [inline default/python/AI/ShipDesignAI.py:381]: Should this be an error?  Is it possible for the cache to be corrupted and still operating correctly?

File: default/python/AI/ShipDesignAI.py [hunk 7]
```diff
@@ -567,9 +562,8 @@ def _update_buildable_items_this_turn(self, verbose=False):
                 if i != idx:
                     available_hulls[i], available_hulls[idx] = available_hulls[idx], available_hulls[i]
             except ValueError:
-                print >> sys.stderr, ("hull in testhull cache not in available_hulls"
-                                      "even though it is supposed to be a proper subset.")
-                traceback.print_exc()
+                warn("hull in testhull cache not in available_hulls even though it is supposed to be a proper subset.",
```
Comment (LGM-Doyle) [inline default/python/AI/ShipDesignAI.py:565]: This should be an error.

File: default/python/AI/ShipDesignAI.py [hunk 8]
```diff
@@ -1043,9 +1037,8 @@ def add_design(self, verbose=False):
                 return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
             except AttributeError:
                 cached_name = Cache.map_reference_design_name[reference_name]
-                print >> sys.stderr, "%s maps to %s in Cache.map_reference_design_name." % (reference_name, cached_name),
-                print >> sys.stderr, "But the design seems not to exist..."
-                traceback.print_exc()
+                warn("%s maps to %s in Cache.map_reference_design_name." % (reference_name, cached_name))
+                warn("But the design seems not to exist...", exc_info=True)
```
Comment (LGM-Doyle) [inline default/python/AI/ShipDesignAI.py:1041]: This should be an error, as it is an event that should never happen.

File: default/python/AI/ShipDesignAI.py [hunk 9]
```diff
@@ -1093,7 +1086,7 @@ def optimize_design(self, additional_parts=(), additional_hulls=(),
         elif isinstance(loc, list):
             planets = loc
         else:
-            print >> sys.stderr, "Invalid loc parameter for optimize_design(). Expected int or list but got", loc
+            warn("Invalid loc parameter for optimize_design(). Expected int or list but got %s" % loc)
```
Comment (LGM-Doyle) [inline default/python/AI/ShipDesignAI.py:1089]: The data type is invalid and this should never happen, so it should be an error.

**Review by LGM-Doyle — COMMENTED** [review]
:smiley: you are using the new logging functions.  

I have quibbled with some of the levels as warnings.  Using error only to indicate unrecoverable errors is too high a threshold for the AI, since if the AI traps an exception and sends empty orders it can be considered as barely operating as intended.

Cases where the event should never happen should be errors and not warnings.  For example a data structure entirely within the python control that is missing a required element, should be an error.  There is no expected/valid reason for the data structure to be missing.  On the other hand a look up table within `fo` missing an id, should only be a warning.  

Another way to consider it is, if the event indicates the algorithm is broken, then it is an error.  If the event indicates a good algorithm has been fed bad data, then it is a warning.

And lastly, an error should indicate something that must be fixed by a developer before release, versus a warning which may be due to bad data/FOCS content.

These are minor concerns, and if you think it better to use `warn()`, this PR is an improvement over the existing code as is.

**Comment by Dilvish-fo** [comment]
@Morlic-fo @Cjkjvfnby Although it would have been much more convenient if LGM-Doyle had gotten his comments in sooner, I think he makes a persuasive argument and this kind of thing is easy enough to change, so let's talk this through a bit more.   I think we should also note a key distinction being that (to my understanding) the error logger will now also always report the error message to the user via chat window.   That seems consistent with his proposed significance for error(), and in any grey area cases it may help decide which type of logger to use.

So, what do you two think about Doyle's proposed distinction between warn and error?

File: default/python/AI/PlanetUtilsAI.py (hunk 2 above)
Comment (Morlic-fo) [reply default/python/AI/PlanetUtilsAI.py:44]: Not even sure why this part of the code exists, this case is handled as soon as the AI code is called. Even if not, we would have a gazillion of errors thrown if an empire does not exist as it is nowhere else in the code.

Probably best to remove that logging completely...

**Comment by Morlic-fo** [comment]
Seems reasonable.

**Comment by Cjkjvfnby** [comment]
> So, what do you two think about Doyle's proposed distinction between warn and error?

Looks ok.

File: default/python/AI/ShipDesignAI.py (hunk 5 above)
Comment (Morlic-fo) [reply default/python/AI/ShipDesignAI.py:373]: This indicates a mismatch between AI cache and server ids. This is as likely an error of the server as it is of the AI code. In my opinion, this should stay on ```warn``` level.

File: default/python/AI/ShipDesignAI.py (hunk 6 above)
Comment (Morlic-fo) [reply default/python/AI/ShipDesignAI.py:381]: The history of this is that some design ids somehow magically changed during turns once (probably related to a moderator ingame). Within a single turn, the cache worked fine. Next turn, there was a mismatch between cached and server ids. So, yes, the cache may very well work fine (though inefficient) if such a mismatch happens.
If there is intra-turn mismatch between the data, there is no way the AI can play anyway as it would get "random" data from the server and pretty much all the code is unlikely to work at that point.

**Comment by Morlic-fo** [comment]
Posted #1746 for the proposed changes that I agree with.

File: default/python/AI/ShipDesignAI.py (hunk 5 above)
Comment (LGM-Doyle) [reply default/python/AI/ShipDesignAI.py:373]: I'd still make it an error, because it is something that a dev needs to fix.  Although maybe not the dev who wrote the python code.

**Comment by Cjkjvfnby** [comment]
@Dilvish-fo, can you summarize it to http://www.freeorion.org/index.php/Python_Development?
